zephyr/lib/regions_mm.c

```diff
@@ -1,53 +1,428 @@
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++) {
+		if (cfg->block_bundles_table[i].block_size > 0)	{
+			total_requested_size += cfg->block_bundles_table[i].block_size *
+				cfg->block_bundles_table[i].number_of_blocks;
+		}
+	}
+
+	if (total_requested_size > new_heap->virtual_region->size)
+		return NULL;
+
+	/* Offset used to calculate where in virtual region to start buffer
+	 * for the memblock
+	 */
+	uint32_t offset = 0;
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++) {
+		if (cfg->block_bundles_table[i].block_size > 0)	{
+			/* Need to create variety of structs and constantly keep theme
+			 * in memory so we need to keep them on sysheap.
+			 */
+			struct sys_mem_blocks *new_allocator =
+				rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+					sizeof(sys_mem_blocks_t));
+			struct sys_bitarray *allocators_bitaray =
+				rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+					sizeof(sys_bitarray_t));
+
+			size_t bitaray_size = SOF_DIV_ROUND_UP(SOF_DIV_ROUND_UP(
+				cfg->block_bundles_table[i].number_of_blocks, 8),
+					sizeof(uint32_t));
+
+			uint32_t *bitaray_bitfield;
+
+			bitaray_bitfield = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+				sizeof(uint32_t) * bitaray_size);
+
+			allocators_bitaray->num_bits =
+				cfg->block_bundles_table[i].number_of_blocks;
+			allocators_bitaray->num_bundles = bitaray_size;
+			allocators_bitaray->bundles = bitaray_bitfield;
+
+			new_allocator->num_blocks =
+				cfg->block_bundles_table[i].number_of_blocks;
+			new_allocator->blk_sz_shift =
+				ilog2(cfg->block_bundles_table[i].block_size);
+			new_allocator->buffer =
+				(uint8_t *)((uint8_t *)new_heap->virtual_region->addr +
+					offset);
+			new_allocator->bitmap = allocators_bitaray;
+
+			offset += cfg->block_bundles_table[i].number_of_blocks *
+				cfg->block_bundles_table[i].block_size;
+			new_heap->physical_blocks_allocators[i] = new_allocator;
+		}
+	}
+
+	new_heap->span = span;
+
+	/* Save the new heap pointer to a linked list */
+	list_item_append(&vm_heaps, &new_heap->node);
+
+	return new_heap;
+}
+
+/**
+ * @brief Alloc function
+ *
+ * Allocates memory on heap from provided heap pointer.
+ * Check if we need to map physical memory for given allocation
+ * and do it when nesscesary.
+ * @param heap pointer to a heap that will be used for allocation.
+ * @param alloc_size Size in bytes to allocate.
+ *
+ * @retval ptr to a new heap if successful
+ * @retval NULL on allocation failure
+ */
+void *vmh_alloc(const struct virtual_memory_heap *heap, size_t alloc_size)
+{
+	void *ptr = NULL;
+	/* Error code we will ghater when allocating on physical block
+	 * allocators.
+	 * Loop will continiue as long as there are blocks to iterate on and
+	 * if allocation keeps returning error code.
+	 * When alloc succeeded it will return 0 on allocation and we break
+	 * form loop and continiue.
+	 */
+	bool allocation_error_code = true;
+	struct active_allocation *new_allocation =
+		rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
```

Comment:
Yes this will be freed in vmh free but this memory has information about allocation not allocation itself. So it has to be stored in sys heap as we can't save it in memory we manage (well we could if we took allocation header aproach but )

Reply:
To quote from alloc.h:
```
 * 1) System Zone. Fixed size heap where alloc always succeeds and is never
 * freed. Used by any init code that will never give up the memory.
```